### PR TITLE
Test: Simplified Kafka K8S test

### PR DESCRIPTION
**Pull request by eloycoto**
- Add a livenessProbe in kafka manifest, wait until kafka is ready.
- Make test code clear
- Add CepGet in kubectl.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

### CODE REVIEW

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

File: test/k8sT/KafkaPolicies.go [hunk 1]
```diff
@@ -107,157 +106,112 @@ var _ = Describe("K8sValidatedKafkaPolicyTest", func() {
...
-		return
-
-		clusterIP, err := kubectl.Get(helpers.DefaultNamespace, "svc").Filter(
-			"{.items[?(@.metadata.name == \"kafka-service\")].spec.clusterIP}")
-		logger.Infof("KafkaPolicyRulesTest: cluster service ip '%s'", clusterIP)
-		Expect(err).Should(BeNil())
-
-		By("Waiting for all Cilium Pods and endpoints to be ready ")
-		By("Waiting for node K8s1")
-		ciliumPod1, _ := kubectl.WaitCiliumEndpointReady(podFilter, helpers.K8s1)
```
Comment (manalibhutiyani) [inline test/k8sT/KafkaPolicies.go]: @eloy: This is a multimode setup, with 

 ```
Kafka multinode setup:
 
 			K8s1:
 				1. empire-backup
 				2. empire-hq
 				3. kafka-broker : ingress policy only
 
 			K8s2:
 			   1. zook
 			   2. empire-outpost-8888
 			   3. empire-outpost-9999
```
 	

You are only testing for k8s2 nodes and have removed the logic for making sure all pods are up on k8s1

File: test/k8sT/KafkaPolicies.go [hunk 2]
```diff
@@ -107,157 +106,112 @@ var _ = Describe("K8sValidatedKafkaPolicyTest", func() {
...
-			}
-			return endpoints1.AreReady()
-		}, "could not get endpoints", &helpers.TimeoutConfig{Timeout: 100})
-
-		Expect(epsStatus1).Should(BeNil())
-
-		endpoints1, err := kubectl.CiliumEndpointsListByLabel(ciliumPod1, podFilter)
-		policyStatus1 := endpoints1.GetPolicyStatus()
-
-		/*
```
Comment (manalibhutiyani) [inline test/k8sT/KafkaPolicies.go]: Why was this comment removed? This helps to understand the Kafka multimode setup. Could you please put it back?

File: test/k8sT/KafkaPolicies.go [hunk 3]
```diff
@@ -107,157 +106,112 @@ var _ = Describe("K8sValidatedKafkaPolicyTest", func() {
...
-
-			K8s2:
-			   1. zook
-			   2. empire-outpost-8888
-			   3. empire-outpost-9999
-		*/
-		By("Testing Kafka endpoint policy enforcement status on K8s1")
-		Expect(policyStatus1[models.EndpointPolicyEnabledNone]).Should(Equal(2))
-		// Only the kafka broker app should have ingress policy enabled.
-		Expect(policyStatus1[models.EndpointPolicyEnabledIngress]).Should(Equal(1))
```
Comment (manalibhutiyani) [inline test/k8sT/KafkaPolicies.go:205]: Why are we not testing the `EndpointPolicyEnabled` status anymore?

File: test/k8sT/KafkaPolicies.go [hunk 4]
```diff
@@ -45,24 +41,23 @@ var _ = Describe("K8sValidatedKafkaPolicyTest", func() {
...
 		logger.Info("Starting")
 		kubectl = helpers.CreateKubectl(helpers.K8s1VMName(), logger)
-		podFilter = "k8s:zgroup=kafkaTestApp"
 
-		kafkaApp = "kafka"
-		zookApp = "zook"
-		backupApp = "empire-backup"
-		empireHqApp = "empire-hq"
-		outpostApp = "empire-outpost"
+		ciliumPod, err = kubectl.GetCiliumPodOnNode(helpers.KubeSystemNamespace, helpers.K8s2)
```
Comment (manalibhutiyani) [inline test/k8sT/KafkaPolicies.go]: Should we not be doing this for k8s1 ? This is a multinode setup?

**Review by manalibhutiyani — CHANGES_REQUESTED** [review]
I think the fact that this is a multi-node setup is not taken into account. Please correct me if I am wrong. Also, the comment explaining the multinode setup helps people to understand the code better when reading it for the first time. Could you put that back.

File: test/k8sT/manifests/kafka-sw-app.yaml [hunk 5]
```diff
@@ -28,6 +28,17 @@ spec:
...
+            port: 9092
+          initialDelaySeconds: 30
+          failureThreshold: 10
+          periodSeconds: 10
+        readinessProbe:
+          tcpSocket:
+            port: 9092
+          initialDelaySeconds: 5
+          periodSeconds: 5
       nodeSelector:
```
Comment (manalibhutiyani) [inline test/k8sT/manifests/kafka-sw-app.yaml]: I think we want to truly randomize it we need to remove the constaints on `nodeSelector:` of all the app pods

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
Hi @manalibhutiyani 

Before you review again, let me explain a few things:

*Delete the nodeselector in manifest*

This is due that we want to schedule pods randomly, and we don't want to have
one pod in each node, with this method all pods are always scheduled in a
random way, so we are sure that different scenarios work,example:

```
vagrant@k8s1:~$ kubectl get pods -o wide -l zgroup=kafkaTestApp
NAME                                   READY     STATUS    RESTARTS   AGE       IP            NODE
empire-backup-67674d68db-rl4jf         1/1       Running   0          58s       10.10.1.238   k8s2
empire-hq-bc4bfbcc6-2v759              1/1       Running   0          58s       10.10.0.4     k8s1
empire-outpost-8888-9b8665f48-4rnpl    1/1       Running   0          58s       10.10.1.147   k8s2
empire-outpost-9999-5d98d9d668-lm2nw   1/1       Running   0          58s       10.10.1.159   k8s2
kafka-broker-7db77df494-wn25m          1/1       Running   1          58s       10.10.1.158   k8s1
zookeeper-9c79c7448-hnd7g              1/1       Running   0          58s       10.10.1.187   k8s2
```

Plus, this is super useful if you want to run the test with only one node instead of two.

*Deleting ciliumPod endpoint list*

As you see we don't check the endpoints status using
`kubectl.CiliumEndpointsListByLabel` and count the output. I added a new method
`kubectl.CepGet` that uses the *Cilium endpoint kubernetes resource* and we can
retrieve that info using `kubectl`. On the other hand, instead of checking how
many endpoints has the policy enabled/disabled, we check that each pod has the
correct value.

*Added a new healthcheck on kafka manifest*

In this case wait until pod is running that Kafka is listening, this helps to
avoid weird race conditions where Pod is `ready` state but kafka is not
listening yet. With this check Kubernetes take care of that.


The PR tried to make things easier, this work will be replicated to
`K8st/Policies.go` and `K8st/Services.go` when this PR get merged, so if you
can review asap I'll be grateful.

Thanks
Eloy

File: test/helpers/kubectl.go [hunk 6]
```diff
@@ -167,6 +167,29 @@ func CreateKubectl(vmName string, log *logrus.Entry) *Kubectl {
...
+// namespaces. If the pod is not present it returns nil
+func (kub *Kubectl) CepGet(namespace string, pod string) *models.Endpoint {
+	log := kub.logger.WithFields(logrus.Fields{
+		"cep":       pod,
+		"namespace": namespace})
+
+	cmd := fmt.Sprintf("%s -n %s get cep %s -o json | jq '.status'", KubectlCmd, namespace, pod)
+	res := kub.Exec(cmd)
+	if !res.WasSuccessful() {
+		log.Debug("cep is not presnet")
```
Comment (manalibhutiyani) [inline test/helpers/kubectl.go]: nit `presnet` -> `present`

File: test/k8sT/KafkaPolicies.go [hunk 7]
```diff
@@ -115,116 +97,86 @@ var _ = Describe("K8sValidatedKafkaPolicyTest", func() {
...
-
 		// We need to produce first, since consumer script waits for
 		// some messages to be already there by the producer.
-		err = kubectl.ExecKafkaPodCmd(
-			helpers.DefaultNamespace, appPods[empireHqApp], fmt.Sprintf(prodHqAnnounce))
-		Expect(err).Should(BeNil(), "Failed to produce to empire-hq on topic empire-announce")
+
+		err := kubectl.ExecKafkaPodCmd(
+			helpers.DefaultNamespace, appPods[empireHqApp], prodHqAnnounce)
+		Expect(err).Should(BeNil(), "EmpireHQApp cannot announce HQ")
```
Comment (manalibhutiyani) [inline test/k8sT/KafkaPolicies.go]: `EmpireHQApp cannot announce HQ` does not convey what Kafka message failed.
Please change this to 
`Failed to produce to empire-hq on topic empire-announce`

File: test/k8sT/KafkaPolicies.go [hunk 8]
```diff
@@ -115,116 +97,86 @@ var _ = Describe("K8sValidatedKafkaPolicyTest", func() {
...
 
 		err = kubectl.ExecKafkaPodCmd(
-			helpers.DefaultNamespace, appPods[outpostApp], fmt.Sprintf(conOutpostAnnoune))
-		Expect(err).Should(BeNil(), "Failed to consume from outpost on topic empire-announce")
+			helpers.DefaultNamespace, appPods[outpostApp], conOutpostAnnoune)
+		Expect(err).Should(BeNil(), "OutpostApp cannot announce Outpost")
```
Comment (manalibhutiyani) [inline test/k8sT/KafkaPolicies.go]: Ditto. this does not convey what Kafka message and topic failed on which action. Please change the message back to `Failed to consume from outpost on topic empire-announce`

File: test/k8sT/KafkaPolicies.go [hunk 9]
```diff
@@ -115,116 +97,86 @@ var _ = Describe("K8sValidatedKafkaPolicyTest", func() {
...
 
 		err = kubectl.ExecKafkaPodCmd(
-			helpers.DefaultNamespace, appPods[empireHqApp], fmt.Sprintf(prodHqDeathStar))
-		Expect(err).Should(BeNil(), "Failed to produce to empire-hq on topic deathstar-plans")
+			helpers.DefaultNamespace, appPods[empireHqApp], prodHqDeathStar)
+		Expect(err).Should(BeNil(), "EmpireHqApp cannot produce messages")
```
Comment (manalibhutiyani) [inline test/k8sT/KafkaPolicies.go]: Ditto:
`EmpireHqApp cannot produce messages` does not convey on which topic. Please change back to
`Failed to produce to empire-hq on topic deathstar-plans`

File: test/k8sT/KafkaPolicies.go [hunk 10]
```diff
@@ -115,116 +97,86 @@ var _ = Describe("K8sValidatedKafkaPolicyTest", func() {
...
 
 		err = kubectl.ExecKafkaPodCmd(
-			helpers.DefaultNamespace, appPods[outpostApp], fmt.Sprintf(conOutDeathStar))
-		Expect(err).Should(BeNil(), "Failed to consume from outpost on topic deathstar-plans")
+			helpers.DefaultNamespace, appPods[outpostApp], conOutDeathStar)
+		Expect(err).Should(BeNil(), "OutpostApp cannot produce messages")
```
Comment (manalibhutiyani) [inline test/k8sT/KafkaPolicies.go]: Ditto. `OutpostApp cannot produce messages` does not convey the topic. Please change back to `Failed to consume from outpost on topic deathstar-plans`

File: test/k8sT/KafkaPolicies.go [hunk 11]
```diff
@@ -115,116 +97,86 @@ var _ = Describe("K8sValidatedKafkaPolicyTest", func() {
...
 
 		err = kubectl.ExecKafkaPodCmd(
-			helpers.DefaultNamespace, appPods[backupApp], fmt.Sprintf(prodBackAnnounce))
-		Expect(err).Should(BeNil(), "Failed to produce to backup on topic empire-announce")
+			helpers.DefaultNamespace, appPods[backupApp], prodBackAnnounce)
+		Expect(err).Should(BeNil(), "BackupApp cannot not announce")
```
Comment (manalibhutiyani) [inline test/k8sT/KafkaPolicies.go]: `BackupApp cannot not announce` ---> `Failed to produce to backup on topic empire-announce`
This specifies exactly what action on what topic failed for the particular apps.

File: test/k8sT/KafkaPolicies.go [hunk 12]
```diff
@@ -115,116 +97,86 @@ var _ = Describe("K8sValidatedKafkaPolicyTest", func() {
...
-		// Only the kafka broker app should have ingress policy enabled.
-		Expect(policyStatus1[models.EndpointPolicyEnabledIngress]).Should(Equal(1))
-		Expect(policyStatus1[models.EndpointPolicyEnabledEgress]).Should(Equal(0))
-		Expect(policyStatus1[models.EndpointPolicyEnabledBoth]).Should(Equal(0))
-
-		By("Testing endpoint policy trace status on kafka-broker node")
-
-		trace := kubectl.CiliumExec(ciliumPod1, fmt.Sprintf(
+			helpers.DefaultNamespace, appPods[outpostApp], prodOutAnnounce)
+		Expect(err).Should(BeNil(), "OutpostApp cannot not annouce")
```
Comment (manalibhutiyani) [inline test/k8sT/KafkaPolicies.go]: `OutpostApp cannot not annouce` does not make a lot of sense to me.
Maybe change to `Failed to produce to outpost on topic empire-announce`

**Review by manalibhutiyani — CHANGES_REQUESTED** [review]

**Comment by eloycoto** [comment]
@manalibhutiyani Sorry, I added the assert messages at that time, and you updated them meanwhile and in the merge I keep the wrong lines. 

This PR should ok now. Sorry!

**Comment by eloycoto** [comment]
test-me-please

**Review by manalibhutiyani — APPROVED** [review]

**Review by ianvernon — APPROVED** [review]

**Pull request merged**